std/functional.d

```diff
@@ -1260,17 +1260,12 @@ Note:
 */
 template memoize(alias fun)
 {
-    import std.traits : ReturnType;
-     // https://issues.dlang.org/show_bug.cgi?id=13580
-    // alias Args = Parameters!fun;
-
-    ReturnType!fun memoize(Parameters!fun args)
+    auto memoize(Args...)(Args args) if (is(typeof(fun(args))))
```

Comment:
This could cause unexpected breakage. This code works right now but won't work after this PR:
``` d
import std.functional : memoize;
static Object objNum(int a) { return new Object(); }
assert(memoize!objNum(0) is memoize!objNum(0U)); 
```

Reply:
That case is now handled, except for the case discussed below, which does not introduce breakage.